File: python/pow_parallel.py
import time
import argparse
import multiprocessing as mp
from blockchain import Block, Blockchain, Transaction, generate_random_transactions
from utils import (
    save_json,
    save_csv,
    print_block_info,
    print_mining_summary,
    format_hash_rate,
    format_time,
    get_config_suffix,
)


def worker_mine(
    worker_id,
    previous_hash,
    transactions_data,
    timestamp,
    difficulty,
    start_nonce,
    step,
    found_flag,
    result_queue,
):
    transactions = [Transaction.from_dict(tx) for tx in transactions_data]
    required_prefix = "0" * difficulty
    nonce = start_nonce

    nonces_tested = 0
    check_interval = 10000

    while True:
        block = Block(previous_hash, transactions, timestamp, nonce)
        nonces_tested += 1
        if block.hash.startswith(required_prefix):
            found_flag.value = 1

            result = {
                "worker_id": worker_id,
                "found": True,
                "nonce": nonce,
                "hash": block.hash,
                "block_data": {
                    "previous_hash": block.previous_hash,
                    "timestamp": block.timestamp,
                    "nonce": block.nonce,
                    "transactions": [tx.to_dict() for tx in block.transactions],
                    "hash": block.hash,
                },
                "nonces_tested": nonces_tested,
            }

            result_queue.put(result)

            return result
        if nonces_tested % check_interval == 0:
            if found_flag.value == 1:
                result = {
                    "worker_id": worker_id,
                    "found": False,
                    "nonces_tested": nonces_tested,
                }
                result_queue.put(result)

                return result
        nonce += step


def mine_block_parallel(pool, previous_hash, transactions, difficulty, num_workers):
    timestamp = time.time()
    transactions_data = [tx.to_dict() for tx in transactions]

    manager = mp.Manager()
    found_flag = manager.Value("i", 0)
    result_queue = manager.Queue()

    async_results = []
    for worker_id in range(num_workers):
        async_result = pool.apply_async(
            worker_mine,
            args=(
                worker_id,
                previous_hash,
                transactions_data,
                timestamp,
                difficulty,
                worker_id,
                num_workers,
                found_flag,
                result_queue,
            ),
        )
        async_results.append(async_result)

    found_block = None
    total_nonces_tested = 0
    worker_stats = {i: 0 for i in range(num_workers)}

    while found_block is None:
        try:
            result = result_queue.get(timeout=0.1)
            worker_id = result["worker_id"]
            nonces_tested = result["nonces_tested"]

            worker_stats[worker_id] += nonces_tested
            total_nonces_tested += nonces_tested

            if result["found"]:
                block_data = result["block_data"]
                block_transactions = [
                    Transaction.from_dict(tx) for tx in block_data["transactions"]
                ]

                found_block = Block(
                    block_data["previous_hash"],
                    block_transactions,
                    block_data["timestamp"],
                    block_data["nonce"],
                )
                found_block.hash = block_data["hash"]

                break
        except:
            pass
    for _ in range(num_workers - 1):
        try:
            result = result_queue.get(timeout=1.0)
            worker_id = result["worker_id"]
            nonces_tested = result["nonces_tested"]

            worker_stats[worker_id] += nonces_tested
            total_nonces_tested += nonces_tested
        except:
            break
    return found_block, total_nonces_tested, worker_stats


def mine_blockchain_parallel(difficulty, num_blocks, num_workers, txs_per_block):
    print(f"Parallel PoW: d={difficulty}, blocks={num_blocks}, workers={num_workers}")

    pool = mp.Pool(processes=num_workers)
    blockchain = Blockchain()

    print(f"Pool created with {num_workers} workers")

    progress_data, performance_metrics, load_balancing_stats = [], [], []
    total_nonces_tested, total_time = 0, 0

    for i in range(num_blocks):
        block_num = i + 1
        transactions = generate_random_transactions(txs_per_block)
        previous_hash = blockchain.get_latest_block().hash

        print(f"[Block {block_num}] Mining...")
        start_time = time.time()
        block, nonces_tested, worker_stats = mine_block_parallel(
            pool, previous_hash, transactions, difficulty, num_workers
        )

        elapsed_time = time.time() - start_time
        hash_rate = nonces_tested / elapsed_time if elapsed_time > 0 else 0

        if blockchain.add_block(block, difficulty):
            print(
                f"  Nonce: {block.nonce}, Time: {format_time(elapsed_time)}, Rate: {format_hash_rate(hash_rate)}"
            )

        for worker_id in sorted(worker_stats.keys()):
            nonces = worker_stats[worker_id]
            percentage = (nonces / nonces_tested * 100) if nonces_tested > 0 else 0
            print(f"    W{worker_id}: {nonces:,} ({percentage:.1f}%)")

        progress_data.append(
            {
                "block_number": block_num,
                "nonce": block.nonce,
                "nonces_tested": nonces_tested,
                "elapsed_time": elapsed_time,
                "cumulative_time": total_time + elapsed_time,
                "hash_rate": hash_rate,
                "hash": block.hash,
                "num_transactions": len(transactions),
            }
        )

        performance_metrics.append(
            {
                "block_number": block_num,
                "difficulty": difficulty,
                "num_workers": num_workers,
                "nonces_tested": nonces_tested,
                "time_seconds": elapsed_time,
                "hash_rate": hash_rate,
                "timestamp": block.timestamp,
                "num_transactions": len(transactions),
            }
        )

        for worker_id, nonces in worker_stats.items():
            load_balancing_stats.append(
                {
                    "block_number": block_num,
                    "worker_id": worker_id,
                    "nonces_tested": nonces,
                    "percentage": (nonces / nonces_tested * 100)
                    if nonces_tested > 0
                    else 0,
                }
            )

        total_nonces_tested += nonces_tested
        total_time += elapsed_time

    pool.close()
    pool.join()
    print("Pool closed")

    print_mining_summary(total_time, total_nonces_tested, num_blocks)

    return {
        "blockchain": blockchain,
        "progress": progress_data,
        "performance": performance_metrics,
        "load_balancing": load_balancing_stats,
        "summary": {
            "total_time": total_time,
            "total_nonces_tested": total_nonces_tested,
            "blocks_mined": num_blocks,
            "average_hash_rate": total_nonces_tested / total_time
            if total_time > 0
            else 0,
            "difficulty": difficulty,
            "num_workers": num_workers,
            "num_blocks": num_blocks,
            "txs_per_block": txs_per_block,
        },
    }


def save_results(results, difficulty, num_blocks, num_workers, txs_per_block):
    config = get_config_suffix(difficulty, num_blocks, txs_per_block, num_workers)
    save_json(
        {"metadata": results["summary"], "blocks": results["progress"]},
        f"pow_mining_parallel_{config}.json",
    )
    save_json(results["blockchain"].to_dict(), f"pow_blockchain_parallel_{config}.json")
    save_csv(results["performance"], f"pow_performance_parallel_{config}.csv")
    save_csv(results["load_balancing"], f"pow_load_balancing_stats_{config}.csv")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-d", "--difficulty", type=int, default=4)
    parser.add_argument("-n", "--num-blocks", type=int, default=5)
    parser.add_argument("-w", "--workers", type=int, default=4)
    parser.add_argument("-t", "--txs-per-block", type=int, default=5)
    args = parser.parse_args()

    try:
        results = mine_blockchain_parallel(
            args.difficulty, args.num_blocks, args.workers, args.txs_per_block
        )
        save_results(
            results, args.difficulty, args.num_blocks, args.workers, args.txs_per_block
        )
        print("FINAL BLOCKCHAIN")
        for idx, block in enumerate(results["blockchain"].chain):
            print_block_info(block, idx)
        print(f"Total: {len(results['blockchain'].chain)} blocks")
    except KeyboardInterrupt:
        print("Interrupted")
    except Exception as e:
        print(f"Error: {e}")
        import traceback

        traceback.print_exc()


if __name__ == "__main__":
    main()

File: rust/src/pow_parallel.rs
use crate::blockchain::{Block, Blockchain};
use crate::utils::{
    MiningProgress, PerformanceMetrics, ThreadPerformance, create_block_data,
    generate_transactions, get_config_suffix, print_progress, save_blockchain,
    save_mining_progress, save_performance_csv, save_thread_performance,
};
use rayon::prelude::*;
use std::error::Error;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::time::Instant;

#[derive(Debug, Clone)]
struct MiningResult {
    block: Block,
    #[allow(dead_code)]
    attempts: u64,
    thread_id: usize,
}

fn mine_block_parallel(
    previous_hash: String,
    timestamp: i64,
    data: String,
    difficulty: usize,
    num_workers: usize,
) -> (Block, u64, Vec<ThreadPerformance>) {
    let found = Arc::new(AtomicBool::new(false));
    let global_attempts = Arc::new(AtomicU64::new(0));

    let result: Arc<parking_lot::Mutex<Option<MiningResult>>> =
        Arc::new(parking_lot::Mutex::new(None));

    let thread_attempts: Vec<Arc<AtomicU64>> = (0..num_workers)
        .map(|_| Arc::new(AtomicU64::new(0)))
        .collect();

    let thread_times: Vec<Arc<parking_lot::Mutex<f64>>> = (0..num_workers)
        .map(|_| Arc::new(parking_lot::Mutex::new(0.0)))
        .collect();

    rayon::ThreadPoolBuilder::new()
        .num_threads(num_workers)
        .build()
        .unwrap()
        .install(|| {
            (0..num_workers).into_par_iter().for_each(|thread_id| {
                let thread_start = Instant::now();
                let mut nonce = thread_id as u64; // every thread starts with a different nonce
                let mut local_attempts = 0u64;

                while !found.load(Ordering::Relaxed) {
                    let block = Block::new(previous_hash.clone(), timestamp, nonce, data.clone());
                    local_attempts += 1;

                    if block.meets_difficulty(difficulty) {
                        // check if it is the first to find it
                        if !found.swap(true, Ordering::SeqCst) {
                            let mining_result = MiningResult {
                                block,
                                attempts: local_attempts,
                                thread_id,
                            };
                            *result.lock() = Some(mining_result);
                        }
                        break;
                    }

                    nonce += num_workers as u64;
                }

                let thread_time = thread_start.elapsed().as_secs_f64();
                thread_attempts[thread_id].fetch_add(local_attempts, Ordering::Relaxed);
                *thread_times[thread_id].lock() = thread_time;
                global_attempts.fetch_add(local_attempts, Ordering::Relaxed);
            });
        });

    let mining_result = result.lock().take().expect("No mining result found");

    let thread_performance: Vec<ThreadPerformance> = (0..num_workers)
        .map(|i| ThreadPerformance {
            thread_id: i,
            blocks_found: if i == mining_result.thread_id { 1 } else { 0 },
            total_attempts: thread_attempts[i].load(Ordering::Relaxed),
            total_time_seconds: *thread_times[i].lock(),
        })
        .collect();

    let total_attempts = global_attempts.load(Ordering::Relaxed);

    (mining_result.block, total_attempts, thread_performance)
}

pub fn run_parallel_mining(
    num_blocks: usize,
    difficulty: usize,
    num_workers: usize,
    transactions_per_block: usize,
) -> Result<(), Box<dyn Error>> {
    println!("=== Parallel ProofOfWork Mining ===");
    println!("Blocks to mine: {}", num_blocks);
    println!("Difficulty: {}", difficulty);
    println!("Number of workers: {}", num_workers);
    println!("Transactions per block: {}", transactions_per_block);
    println!();

    let start_time = Instant::now();
    let mut blockchain = Blockchain::new(difficulty);
    let mut mining_progress = Vec::new();
    let mut total_attempts = 0u64;
    let mut all_thread_performance = Vec::new();

    for i in 1..=num_blocks {
        println!("Mining block {} with {} workers...", i, num_workers);
        let block_start = Instant::now();

        let transactions = generate_transactions(i, transactions_per_block);
        let block_data = create_block_data(i, &transactions);

        let previous_hash = blockchain.last_block().hash.clone();
        let timestamp = chrono::Utc::now().timestamp();

        let (block, attempts, thread_perf) = mine_block_parallel(
            previous_hash,
            timestamp,
            block_data,
            difficulty,
            num_workers,
        );
        let block_time = block_start.elapsed().as_secs_f64();

        total_attempts += attempts;

        let progress = MiningProgress {
            block_number: i,
            nonce: block.nonce,
            hash: block.hash.clone(),
            nonces_tested: attempts,
            time_seconds: block_time,
        };
        mining_progress.push(progress);

        for perf in thread_perf {
            all_thread_performance.push(perf);
        }

        let winner_thread = block.nonce % num_workers as u64;
        println!(
            "  Found by thread {}, Nonce: {} | Hash: {}... | Attempts: {} | Time: {:.3}s",
            winner_thread,
            block.nonce,
            &block.hash[..16],
            attempts,
            block_time
        );

        blockchain.add_block(block);

        let elapsed = start_time.elapsed().as_secs_f64();
        print_progress(i, num_blocks, total_attempts, elapsed);
        println!();
    }

    let total_time = start_time.elapsed().as_secs_f64();
    let hash_rate = total_attempts as f64 / total_time;
    let avg_time_per_block = total_time / num_blocks as f64;

    println!("=== Mining Complete ===");
    println!("Total blocks mined: {}", num_blocks);
    println!("Total time: {:.3}s", total_time);
    println!("Total attempts: {}", total_attempts);
    println!("Hash rate: {:.2} H/s", hash_rate);
    println!("Average time per block: {:.3}s", avg_time_per_block);
    println!("Blockchain valid: {}", blockchain.is_valid());
    println!();

    println!("=== Thread Performance ===");
    let mut thread_stats: Vec<(usize, u64, usize)> = Vec::new();
    for thread_id in 0..num_workers {
        let thread_data: Vec<&ThreadPerformance> = all_thread_performance
            .iter()
            .filter(|p| p.thread_id == thread_id)
            .collect();

        let total_thread_attempts: u64 = thread_data.iter().map(|p| p.total_attempts).sum();
        let blocks_found: usize = thread_data.iter().map(|p| p.blocks_found).sum();

        thread_stats.push((thread_id, total_thread_attempts, blocks_found));
    }

    for (thread_id, attempts, blocks_found) in thread_stats {
        let percentage = (attempts as f64 / total_attempts as f64) * 100.0;
        println!(
            "Thread {}: {} attempts ({:.2}%), {} blocks found",
            thread_id, attempts, percentage, blocks_found
        );
    }
    println!();

    let config_suffix = get_config_suffix(
        difficulty,
        num_blocks,
        Some(transactions_per_block),
        Some(num_workers),
    );

    println!("Saving results...");
    let mining_filename = format!("pow_mining_parallel_rust_{}.json", config_suffix);
    let blockchain_filename = format!("pow_blockchain_parallel_rust_{}.json", config_suffix);
    let performance_filename = format!("pow_performance_parallel_rust_{}.csv", config_suffix);
    let thread_filename = format!("pow_thread_performance_rust_{}.json", config_suffix);

    save_mining_progress(&mining_progress, &mining_filename)?;
    save_blockchain(&blockchain, &blockchain_filename)?;
    save_thread_performance(&all_thread_performance, &thread_filename)?;

    let metrics = PerformanceMetrics {
        total_blocks: num_blocks,
        difficulty,
        total_time_seconds: total_time,
        total_nonces_tested: total_attempts,
        hash_rate,
        //avg_time_per_block,
    };
    save_performance_csv(&metrics, &performance_filename)?;

    println!("Results saved to output/");
    println!("  - {}", mining_filename);
    println!("  - {}", blockchain_filename);
    println!("  - {}", performance_filename);
    println!("  - {}", thread_filename);

    Ok(())
}
